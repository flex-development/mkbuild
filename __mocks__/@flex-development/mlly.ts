/**
 * @file Mocks - @flex-development/mlly
 * @module mocks/flex-development/mlly
 * @see https://github.com/flex-development/mlly
 */

/**
 * [`@flex-development/mlly`][1] module type.
 *
 * [1]: https://github.com/flex-development/mlly
 */
type Actual = typeof import('@flex-development/mlly')

/**
 * `@flex-development/mlly` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('@flex-development/mlly')

export const CONDITIONS = actual.CONDITIONS
export const RESOLVE_EXTENSIONS = actual.RESOLVE_EXTENSIONS
export const fillModules = vi.fn(actual.fillModules)
export const findStaticImports = vi.fn(actual.findStaticImports)
export const getSource = vi.fn(actual.getSource)
export const readPackageJson = vi.fn(actual.readPackageJson)
export const resolveModule = vi.fn(actual.resolveModule)
export const resolveModules = vi.fn(actual.resolveModules)
export const toDataURL = vi.fn(actual.toDataURL)
export const toURL = vi.fn(actual.toURL)
