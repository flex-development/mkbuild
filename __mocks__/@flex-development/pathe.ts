/**
 * @file Mocks - @flex-development/pathe
 * @module mocks/flex-development/pathe
 * @see https://github.com/flex-development/pathe
 */

/**
 * [`@flex-development/pathe`][1] module type.
 *
 * [1]: https://github.com/flex-development/pathe
 */
type Actual = typeof import('@flex-development/pathe')

/**
 * `@flex-development/pathe` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('@flex-development/pathe')

export const basename = vi.fn(actual.basename)
export const dirname = vi.fn(actual.dirname)
export const extname = vi.fn(actual.extname)
export const format = vi.fn(actual.format)
export const join = vi.fn(actual.join)
export const parse = vi.fn(actual.parse)
export const relative = vi.fn(actual.relative)
export const resolve = vi.fn(actual.resolve)

export default {
  basename,
  dirname,
  extname,
  format,
  join,
  parse,
  relative,
  resolve
}
