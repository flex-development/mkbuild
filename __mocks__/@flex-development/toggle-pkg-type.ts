/**
 * @file Mocks - @flex-development/toggle-pkg-type
 * @module mocks/flex-development/toggle-pkg-type
 * @see https://github.com/flex-development/toggle-pkg-type
 */

export default vi.fn()
