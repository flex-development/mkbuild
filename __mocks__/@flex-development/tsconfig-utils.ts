/**
 * @file Mocks - @flex-development/tsconfig-utils
 * @module mocks/flex-development/tsconfig-utils
 * @see https://github.com/flex-development/tsconfig-utils
 */

/**
 * [`@flex-development/tsconfig-utils`][1] module type.
 *
 * [1]: https://github.com/flex-development/tsconfig-utils
 */
type Actual = typeof import('@flex-development/tsconfig-utils')

/**
 * `@flex-development/tsconfig-utils` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>(
  '@flex-development/tsconfig-utils'
)

export const ScriptTarget = actual.ScriptTarget
export const normalizeCompilerOptions = vi.fn(actual.normalizeCompilerOptions)
export const normalizeModuleResolution = vi.fn(actual.normalizeModuleResolution)
export const resolvePaths = vi.fn(actual.resolvePaths)
