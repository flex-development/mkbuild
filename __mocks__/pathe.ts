/**
 * @file Mocks - pathe
 * @module mocks/pathe
 * @see https://github.com/unjs/pathe
 */

import path, { type FormatInputPathObject } from 'node:path'

/**
 * Return the last portion of `p`.
 *
 * @param {string} p - Path to evaluate
 * @return {string} File name
 */
export const basename = vi.fn((p: string): string => path.basename(p))

/**
 * Returns the directory name of `p`.
 *
 * @param {string} p - Path to evaluate
 * @return {string} Directory containing `p`
 */
export const dirname = vi.fn((p: string): string => path.dirname(p))

/**
 * Returns the extension of `p`, from the last `'.'` to end of string in the
 * last portion of the path.
 *
 * If there is no `'.'` in the last portion of the path or the first character
 * of it is `'.'`, an empty string will be returned.
 *
 * @param {string} p - Path to evaluate
 * @return {string} Extension of `p`
 */
export const extname = vi.fn((p: string) => path.extname(p))

/**
 * Returns a path string from an object - the opposite of {@link parse}.
 *
 * @param {FormatInputPathObject} p - Path object to evaluate
 * @return {path.ParsedPath} Object representing `p` parsed
 */
export const format = vi.fn((p: FormatInputPathObject) => path.format(p))

/**
 * Parses `p`.
 *
 * @param {string} p - Path to evaluate
 * @return {path.ParsedPath} Object representing `p` parsed
 */
export const parse = vi.fn((p: string) => path.parse(p))

/**
 * Returns the relative path from `from` to `to` based on the current working
 * directory.
 *
 * @param {string} from - Start path
 * @param {string} to - Destination path
 * @return {string} Relative path from `from` to `to`
 */
export const relative = vi.fn((from: string, to: string): string => {
  return path.relative(from, to)
})

/**
 * Starting from leftmost {from} parameter, resolves {to} to an absolute path.
 *
 * @param {string[]} paths - Path segments
 * @return {string} Absolute path composed from `paths`
 */
export const resolve = vi.fn((...paths: string[]): string => {
  return path.resolve(...paths)
})
