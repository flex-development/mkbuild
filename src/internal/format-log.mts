/**
 * @file Internal - formatLog
 * @module mkbuild/internal/formatLog
 */

import pathe from '@flex-development/pathe'
import { ksort } from '@flex-development/tutils'
import type { LogLevel, RollupLog } from 'rollup'

/**
 * Format a rollup `log`.
 *
 * @see https://rollupjs.org/plugin-development/#onlog
 * @see https://github.com/rollup/rollup/issues/5699
 * @see {@linkcode LogLevel}
 * @see {@linkcode RollupLog}
 *
 * @internal
 *
 * @this {void}
 *
 * @param {LogLevel | 'error'} level
 *  Rollup log level
 * @param {RollupLog} log
 *  Rollup log object
 * @return {undefined}
 */
function formatLog(
  this: void,
  level: LogLevel | 'error',
  log: RollupLog
): undefined {
  /**
   * Module id.
   *
   * @const {string | undefined} id
   */
  const id: string | undefined = log.id || log.loc?.file

  if (log.plugin) {
    log.message = log.message.slice(`[plugin ${log.plugin}]`.length + 1)
  }

  if (id) {
    /**
     * String to search for in {@linkcode log.message}.
     *
     * @var {string} seed
     */
    let seed: string = pathe.basename(id)

    if (log.loc) seed += ` (${log.loc.line}:${log.loc.column}):`

    log.message = log.message.slice(log.message.indexOf(seed) + seed.length + 1)
    log.message = log.message.trimStart()
  }

  log.type ??= log.level ??= level
  return void ksort(log)
}

export default formatLog
