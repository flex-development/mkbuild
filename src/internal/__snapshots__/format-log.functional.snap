// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`functional:internal/formatLog > should format rollup \`log\` (0) 1`] = `
{
  "code": "EMPTY_BUNDLE",
  "level": "warn",
  "message": "Generated an empty chunk: "interfaces/task".",
  "names": [
    "interfaces/task",
  ],
  "type": "warn",
}
`;

exports[`functional:internal/formatLog > should format rollup \`log\` (1) 1`] = `
{
  "code": "PLUGIN_WARNING",
  "hook": "renderStart",
  "id": \${process.cwd()}/tsconfig.build.json,
  "level": "warn",
  "message": "Compiler option 'sourceMap' requires a value of type boolean.",
  "meta": {
    "category": 1,
  },
  "plugin": "mkbuild:dts",
  "pluginCode": 5024,
  "type": "warn",
}
`;

exports[`functional:internal/formatLog > should format rollup \`log\` (2) 1`] = `
{
  "code": "PLUGIN_WARNING",
  "frame": "        : (await import(url.href) as { default?: Config | null }).default",
  "hook": "transform",
  "level": "warn",
  "loc": {
    "column": 18,
    "file": \${process.cwd()}/src/utils/load-build-config.mts,
    "line": 96,
  },
  "message": "This "import" expression will not be bundled because the argument is not a string literal",
  "meta": {
    "level": "warning",
    "notes": [],
  },
  "plugin": "mkbuild:esbuild",
  "pluginCode": "unsupported-dynamic-import",
  "pos": 2188,
  "type": "warn",
}
`;
